Support autograd's numpy with lambdify

RJTK opened the pull request:
In relation to #14686 I have added functionality and tests for lambdify to work with a wrapped version of numpy.

In order for [autograd](https://github.com/HIPS/autograd) to differentiate code it needs to be implemented with autograd's thinly wrapped numpy version.  In my use case, I have used SymPy to symbolically evaluate some integrals, then generated numeric functions with lambdify.  These functions are used and composed with plain numpy throughout a number of additional steps until finally I need a gradient to pass into scipy.optimize.  Since SymPy lambdifies to plain numpy (rather than autograd's wrapped version) I can't directly use autograd to compute the needed gradient.  This patch is intended to fix that problem.

It is in principle possible to use SymPy's symbolic diff methods to supply autograd with the necessary derivatives, however, if the user wanted higher order derivatives, or are using a lot of SymPy lambdified functions, more and more symbolic derivatives need to be added.  Since SymPy's numpy printer already does most of the work, it seems fairly straightforward to allow lambdifying to autograd's numpy. 

Usage example:

```
    from sympy import symbols, Matrix, lambdify
    from autograd import numpy as anp
    from autograd import jacobian

    x, y, z = symbols('x, y, z')
    g = Matrix([x * y, y**2 * z])
    g_lmbda = lambdify((x, y, z), g, modules="autograd.numpy")

    # Need a wrapper so that the argument is a single vector (is there a better way?)
    def gv_lmbda(xx):
        return g_lmbda(*xx)

    # Define another function that makes use of the SymPy generated code
    def h(xx):
        return gv_lmbda(xx)**2 + anp.log(gv_lmbda(xx))

    h_jac = jacobian(h)

    In [351]: h_jac(anp.array([1., 2., 3.]))
    Out[351]: 
    array([[[  9.        ,   4.5       ,   0.        ]],

           [[  0.        , 289.        ,  96.33333333]]])
```

<!-- BEGIN RELEASE NOTES -->
* utilities
    * lambdify can now generate functions compatible with autograd.
<!-- END RELEASE NOTES -->

RJTK committed:
Removed ',' after '...' which caused a doctest to fail.

RJTK committed:
lambdify: Updated to support using moduels="autograd.numpy"

autograd (https://github.com/HIPS/autograd) provides methods
for differentiating (almost) arbitrary numpy code.  However,
it requires functions to be implemented with it's own thinly
wrapped version of numpy (autograd.numpy).  This patch adds
support and tests for lambdifying with this version of numpy
and thus allows one to use autograd on functions that are
lambdified with SymPy.

The use case for this feature is where you have autograd.numpy
code mixed in with functions lambdified by SymPy.  It is
possible to provide gradients to autograd that are lambdified
from sympy's symbolic diff, however, computing many gradients
or higher order gradients would require the user to supply more
and more lambdified symbolic gradients.

RJTK committed:
lambdify: Updated to support using modules="autograd.numpy"

In order for autograd (https://github.com/HIPS/autograd) to
compute gradients of numpy code, it needs to have been
generated by a thinly wrapped version of numpy: autograd.numpy.
It may be desirable to compute derivatives with autograd,
rather than using sympybol differentiation in SymPy if
a lambdified SymPy function is only one small part of a
function implemented mostly with numpy operations.

RJTK committed:
lambdify: Updated to support using moduels="autograd.numpy"

autograd (https://github.com/HIPS/autograd) provides methods
for differentiating (almost) arbitrary numpy code.  However,
it requires functions to be implemented with it's own thinly
wrapped version of numpy (autograd.numpy).  This patch adds
support and tests for lambdifying with this version of numpy
and thus allows one to use autograd on functions that are
lambdified with SymPy.

The use case for this feature is where you have autograd.numpy
code mixed in with functions lambdified by SymPy.  It is
possible to provide gradients to autograd that are lambdified
from sympy's symbolic diff, however, computing many gradients
or higher order gradients would require the user to supply more
and more lambdified symbolic gradients.

RJTK committed:
lambdify: Updated to support using moduels="autograd.numpy"

autograd (https://github.com/HIPS/autograd) provides methods
for differentiating (almost) arbitrary numpy code.  However,
it requires functions to be implemented with it's own thinly
wrapped version of numpy (autograd.numpy).  This patch adds
support and tests for lambdifying with this version of numpy
and thus allows one to use autograd on functions that are
lambdified with SymPy.

The use case for this feature is where you have autograd.numpy
code mixed in with functions lambdified by SymPy.  It is
possible to provide gradients to autograd that are lambdified
from sympy's symbolic diff, however, computing many gradients
or higher order gradients would require the user to supply more
and more lambdified symbolic gradients.

RJTK committed:
Merge branch '14686_lambdify_autograd_numpy' of https://github.com/RJTK/sympy into 14686_lambdify_autograd_numpy

RJTK committed:
Merge branch '14686_lambdify_autograd_numpy' of https://github.com/RJTK/sympy into 14686_lambdify_autograd_numpy

Conflicts:
	sympy/utilities/tests/test_lambdify.py

RJTK commented:
The failed check seems to be caused by a network problem?

```
3.5 is not installed; attempting download
Downloading archive: https://s3.amazonaws.com/travis-python-archives/binaries/ubuntu/14.04/x86_64/python-3.5.tar.bz2
$ curl -sSf -o python-3.5.tar.bz2 ${archive_url}
curl: (7) Failed to connect to s3.amazonaws.com port 443: Connection timed out
Unable to download 3.5 archive. The archive may not exist. Please consider a different version.
```
Should the check just be retried?

asmeurer commented on sympy/utilities/tests/test_lambdify.py:
```diff
@@ -361,6 +493,20 @@ def test_numpy_inverse():
+    # NOTE: A**-1 creates a symbolic inverse -- it doesn't map to np.linalg.inv
+    # there is thus no reason to test the jacobian since it's the same as
+    # differentiating any regular matrix.
+    return
```
These empty `return` statements are not needed.

asmeurer commented:
Can you add `autograd` to the optional dependency list in Travis?

RJTK committed:
Added autograd to .travis.yml

RJTK committed:
Removed empty return statements

asmeurer commented:
There is a test failure in Python 2.

RJTK committed:
external/importtools: Added a less restrictive version comparison for versions like 1.1.post1

RJTK committed:
lambdify: Fixed bug in py2.7 and suppressed an anticipated warning

asmeurer commented:
What is the motivation behind the importtools changes?

RJTK commented:
The function `StrictVersion` will throw errors unless the version number is made up only of period separated integers.  However, it is possible to have version numbers e.g. '1.1.1a', '1.1.post1' ...  The version number of Cython in the python2.7 environment for my system is '0.20.1.post0', which caused an exception to be raised by `StrictVersion`.

I believe these are relevant: [PEP386](https://www.python.org/dev/peps/pep-0386/) which is now superseded by [PEP440](https://www.python.org/dev/peps/pep-0440/).

The code I added preserves the old functionality with `StrictVersion`, and tries to deal with an exception with `parse_version`.

moorepants commented:
@RJTK Can you add release notes to the first comment box here as per: https://github.com/sympy/sympy/wiki/Writing-Release-Notes

moorepants commented:
Also, please resolve conflicts with the master branch.

moorepants commented on sympy/external/importtools.py:
```diff
@@ -157,7 +157,29 @@ def import_module(module, min_module_version=None, min_python_version=None,
+                break
+            except ImportError:  # could pkg_resources be missing?
+                warnings.warn("pkg_resources is not available.",
+                              UserWarning)
```
setuptools is not a required dependency of sympy. No need to warn a user if it isn't there.

RJTK replied:
Addressed this.  I don't remember why I put the warning there.

moorepants reviewed (changes requested):
This looks great. The `importtools` change is a bigger issue. Maybe open a new issue and solve it separately? Also, I'd like to see some form of documentation for this. Maybe some examples in lambdify's docstring.

RJTK committed:
Merge branch 'master' of https://github.com/sympy/sympy

RJTK committed:
Remove warnings from importtools.

RJTK committed:
Added a blurb to the lambdify docstring.

RJTK commented:
I added some info to the docstring.  I think this feature has a pretty specialized use case, since if you are using Sympy you would probably just want to obtain a symbolic gradient?

I implemented this feature beause I wound up with a rather complicated expression from some Gaussian Process models (see http://mlg.eng.cam.ac.uk/pub/pdf/Saa11.pdf -- I used Sympy to implement the symbolic computations for eq 2.46 - 2.49).  I don't remember exactly the circumstances, but probably Sympy's symbolic gradient was taking too long, or I for some other reason couldn't get it to work for my problem.

I have only used Sympy for this project, so it would be nice to hear if other people would find it useful to integrate it with autograd.

RJTK commented:
In regards the modification to importtools, I believe it only has an impact in some edge cases where it will catch an exception and try to use parse_version.  As far as I understand it preserves the old functionality in cases where an exception wouldn't be thrown?

RJTK committed:
Resolve conflicts.

asmeurer commented on sympy/utilities/lambdify.py:
```diff
@@ -184,8 +189,8 @@ def lambdify(args, expr, modules=None, printer=None, use_imps=True,
 
-     - the strings "math", "mpmath", "numpy", "numexpr", "scipy", "sympy",
-       "tensorflow"
+     - the strings "math", "mpmath", "numpy", "autogra.numpy", "numexpr",
```
```suggestion
     - the strings "math", "mpmath", "numpy", "autograd.numpy", "numexpr",
```

asmeurer commented:
There are test failures. Looks like the printer will need to be modified to print `autograd.numpy` instead of `numpy`. Alternately, it may be a good idea to add `from autograd import numpy` to the imports. There is also a Sphinx error in the docstring.

Otherwise, this looks fine. This will create some minor merge conflicts with my PR https://github.com/sympy/sympy/pull/13485, depending on which is merged first.

asmeurer commented on sympy/utilities/lambdify.py:
```diff
@@ -184,8 +189,8 @@ def lambdify(args, expr, modules=None, printer=None, use_imps=True,
-     - the strings "math", "mpmath", "numpy", "numexpr", "scipy", "sympy",
-       "tensorflow"
+     - the strings "math", "mpmath", "numpy", "autogra.numpy", "numexpr",
+     "scipy", "sympy", "tensorflow"
```
I think the Sphinx error is because this line needs to be indented. 
```suggestion
       "scipy", "sympy", "tensorflow"
```

asmeurer commented on sympy/utilities/lambdify.py:
```diff
@@ -210,6 +215,12 @@ def lambdify(args, expr, modules=None, printer=None, use_imps=True,
 
+    Using autograd.numpy is only necessary if you intend to use the
+    resulting function with the autograd module for automatic
+    differentiation.  This is useful for instance if Sympy is used only
```
```suggestion
    differentiation.  This is useful for instance if SymPy is used only
```

czgdp1807 commented:
@RJTK Please resolve the conflicts and apply the suggestions.
